Add config marshaler.

jefchien opened the pull request:
**Description:**
Add `Marshal` function to `confmap`. Update types implementing `encoding.TextUnmarshaler` to implement `encoding.TextMarshaler` as well. Add a mapstructure encoder that encodes the config provided using the existing mapstructure tags.

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector/issues/5418

**Testing:** Added unit test to validate the unmarshal/marshal cycle.

bogdandrutu reviewed (commented):
Forgot to press submit, sorry.

jefchien commented on config/configtelemetry/configtelemetry.go:
```diff
@@ -57,6 +58,11 @@ func (l Level) String() string {
 	return "unknown"
 }
 
+// MarshalText marshals Level to text.
+func (l Level) MarshalText() (text []byte, err error) {
+	return []byte(l.String()), nil
+}
```
This one is needed because `UnmarshalText` expects a string value matching one of the `Level` strings. `Level` is an int32, so without the `MarshalText` function, the marshaler would marshal it as an int which would break the marshal/unmarshal cycle.

jefchien commented on config/configcompression/compressionType.go:
```diff
@@ -32,6 +32,10 @@ func IsCompressed(compressionType CompressionType) bool {
 	return compressionType != empty && compressionType != none
 }
 
+func (ct CompressionType) MarshalText() (out []byte, err error) {
+	return []byte(ct), nil
+}
```
You're right. This one isn't really necessary since the `CompressionType` is a string, so would be unmarshalled correctly as is.

jefchien commented on config/identifiable.go:
```diff
@@ -69,6 +69,11 @@ func (id ComponentID) Name() string {
 	return id.nameVal
 }
 
+// MarshalText implements the encoding.TextMarshaler interface.
+func (id ComponentID) MarshalText() (text []byte, err error) {
+	return []byte(id.String()), nil
+}
```
Without the `MarshalText` function, the `ComponentID` would be considered a struct and be unmarshalled using that logic and would not be usable as a map key.

jefchien commented on receiver/otlpreceiver/config.go:
```diff
@@ -82,3 +83,7 @@ func (cfg *Config) Unmarshal(componentParser *confmap.Conf) error {
 	return nil
 }
+
+func (cfg *Config) Marshal(enc configencoder.Encoder) (interface{}, error) {
+	return enc.Encode(cfg)
+}
```
This is required because I added the `Marshal` function to the `Unmarshallable` interface to force configs that implement `Unmarshal` to implement the `Marshal` function. In the case of the otlp receiver, it can be encoded as is since the `Unmarshal` function simply nils out unset protocols, which can be retained when marshalling with `omitempty`.

jefchien replied:
Separated the `Marshal` function into a `Marshallable` interface.

jefchien commented on config/common.go:
```diff
@@ -32,6 +33,9 @@ type Unmarshallable interface {
 	Unmarshal(component *confmap.Conf) error
+	// Marshal is a function that marshals the config into an interface to be encoded by
+	// the YAML encoder. Provides the configencoder.Encoder.
+	Marshal(encoder configencoder.Encoder) (interface{}, error)
```
Changed it to marshal into a `confmap.Conf`.

jefchien commented:
@bogdandrutu PTAL when you get the chance. Refactored the marshaler to be more similar to the unmarshaler.

bogdandrutu commented on config/common.go:
```diff
@@ -34,6 +37,14 @@ type Unmarshallable interface {
+type Marshallable interface {
+	// Marshal is a function that marshals the config into an interface to be encoded by
+	// the YAML encoder. Provides the configencoder.Encoder.
+	Marshal(encoder configencoder.Encoder) (*confmap.Conf, error)
```
Why not passing a `*confmap.Conf` and the idea is to "Marshal into the Conf"? Why is an encoder needed?

jefchien replied:
The encoder is there to assist the user with marshaling their config. It gives the config a starting point to marshal the config and is optional to use.
```
func (t TestMarshallableExtension) Marshal(enc configencoder.Encoder) (*confmap.Conf, error) {
	out, err := enc.Encode(t)
	if err != nil {
		return nil, err
	}
	data, ok := out.(map[string]interface{})
	if !ok {
		return nil, errors.New("unable to encode struct")
	}
	conf := confmap.NewFromStringMap(data)
	conf.Delete("omit")
	return conf, nil
}
```
Right now, the marshaler doesn't encode the string map it receives from `confmap.Conf` after calling this function. Maybe it makes more sense to allow the config to just marshal into the `confmap.Conf` and then encode that result in the marshaler after.

The `confmap.Conf` doesn't have many functions enabling a user to change the underlying map, so I imagine the user would need to make the changes before creating the `confmap.Conf` with `NewFromStringMap`.

jefchien replied:
@bogdandrutu I'm not opposed to passing in a `*confmap.Conf` instead, but how could a user marshal into it? It doesn't look like the underlying `koanf` instance provides any functions for modification aside from `Merge`/`Delete`.

jefchien commented:
Changed the `Marshallable` interface to take in a `confmap.Conf`. Added a function on the `confmap.Conf` that uses the encoder and uses `koanf.Merge` to "marshal" it into the `confmap.Conf`.

bogdandrutu commented on service/configmarshaler/defaultmarshaler.go:
```diff
@@ -0,0 +1,158 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package configmarshaler // import "go.opentelemetry.io/collector/service/configmarshaler"
```
Why do you need this to be public?

jefchien replied:
The main use case I see for making this public is for downstream distributions or other dependent projects to be able to use this when trying to convert between an existing configuration and the OTEL configuration. In some cases, it may be easier for users to programmatically translate and generate complex configurations. This public marshaler provides a way for them to marshal the generated configuration into a YAML file, which can then be used at runtime.

bogdandrutu replied:
I would split this in the next PR then. I have ideas on what we can do to avoid this.

jefchien replied:
Okay. Removed the defaultmarshaler from this PR.

jefchien commented on confmap/confmap.go:
```diff
@@ -93,6 +108,11 @@ func (l *Conf) Merge(in *Conf) error {
 	return l.k.Merge(in.k)
 }
 
+// Delete deletes the value at the given key if present.
+func (l *Conf) Delete(key string) {
+	l.k.Delete(key)
+}
```
You're right. This isn't necessary. Can remove it.

bogdandrutu commented on config/exporter.go:
```diff
@@ -33,6 +33,13 @@ func UnmarshalExporter(conf *confmap.Conf, cfg Exporter) error {
 	return unmarshal(conf, cfg)
 }
 
+// MarshalExporter helper function to marshal an Exporter config.
+// It checks if the config implements Marshallable and uses that if available,
+// otherwise uses Map.Marshal.
+func MarshalExporter(conf *confmap.Conf, cfg Exporter) error {
+	return marshal(conf, cfg)
+}
```
Let's not add the `Marshal[Component]` for the moment, I think we can make `confmap.Marshal` smart enough to check for that interface and automatically call it if present.

bogdandrutu replied:
Here is one of my previous try to do the same with `Unmarshal` https://github.com/open-telemetry/opentelemetry-collector/pull/4879/files

jefchien replied:
Okay. Removed it and changed the encoder to accept decode hooks. Added a hook to detect the interface.

bogdandrutu commented on config/common.go:
```diff
@@ -34,6 +34,14 @@ type Unmarshallable interface {
 
+// Marshallable defines an optional interface for custom configuration marshalling.
+// A configuration struct can implement this interface to override the default marshalling.
+type Marshallable interface {
```
Let's have this in `confmap`, I think in a followup PR we can have the `Unmarshallable` also moved to confmap and deprecated from here.

jefchien replied:
Moved it under `confmap`.

bogdandrutu reviewed (commented):
Sorry for the delay, I am all focus on this PR now :)

bogdandrutu commented on config/identifiable.go:
```diff
@@ -69,6 +69,11 @@ func (id ComponentID) Name() string {
 	return id.nameVal
 }
 
+// MarshalText implements the encoding.TextMarshaler interface.
+func (id ComponentID) MarshalText() (text []byte, err error) {
+	return []byte(id.String()), nil
+}
+
```
Is this needed?

jefchien replied:
This is needed because otherwise the `ComponentID` will encode as a struct, which will break the marshalling.

bogdandrutu replied:
you are right :) Can you add a comment that this marshals the 2 values as one string bla bla bla :))

bogdandrutu commented on confmap/marshaler.go:
```diff
@@ -0,0 +1,24 @@
+// Marshaler defines an optional interface for custom configuration marshalling.
+// A configuration struct can implement this interface to override the default
+// marshalling.
+type Marshaler interface {
+	// Marshal the config into a Conf in a custom way.
+	// The Conf will be empty and can be merged into.
+	Marshal(component *Conf) error
+}
```
We can have this in the same file as unmarshaler.

jefchien replied:
It's been moved into the same confmap file.

bogdandrutu reviewed (commented):
Please explain why the Encoder is still needed.

jefchien commented:
The Encoder handles all of the reflection needed to respect the mapstructure tags and is used by confmap to marshal. Without it, the marshal/unmarshal cycle would break and the configs being generated would be invalid.

bogdandrutu commented:
> The Encoder handles all of the reflection needed to respect the mapstructure tags and is used by confmap to marshal. Without it, the marshal/unmarshal cycle would break and the configs being generated would be invalid.

Then the dependency graph is totally unexpected. Move that to `confmap/internal/mapstructure` if needed.

jefchien commented:
> > The Encoder handles all of the reflection needed to respect the mapstructure tags and is used by confmap to marshal. Without it, the marshal/unmarshal cycle would break and the configs being generated would be invalid.
> 
> Then the dependency graph is totally unexpected. Move that to `confmap/internal/mapstructure` if needed.

Moved under `confmap/internal/mapstructure`. The old path was a holdover from the initial implementation where the confmap didn't handle the marshalling.

bogdandrutu commented on receiver/otlpreceiver/config.go:
```diff
@@ -32,8 +32,8 @@ const (
 type Protocols struct {
-	GRPC *configgrpc.GRPCServerSettings `mapstructure:"grpc"`
-	HTTP *confighttp.HTTPServerSettings `mapstructure:"http"`
+	GRPC *configgrpc.GRPCServerSettings `mapstructure:"grpc,omitempty"`
+	HTTP *confighttp.HTTPServerSettings `mapstructure:"http,omitempty"`
```
do we really need omitempty on pointers? Also why do we need omitempty in this case, and in general, is not documented.

jefchien replied:
If `omitempty` isn't included, the marshalled config will include a `grpc` or `http` field even for nil pointers. It'll marshal as `grpc: {}` or `http: {}`, which is fine and valid, but unnecessarily clutters up the config. Don't see a point in including the field if it isn't being used and isn't set.

bogdandrutu replied:
let's not add more API surface if not needed. I would avoid having `omitempty ` until we are clear that we need it.

bogdandrutu replied:
Ping on this.

jefchien replied:
Sorry. Must've forgotten to remove this.

bogdandrutu commented on service/telemetry/config.go:
```diff
@@ -88,7 +88,7 @@ type LogsConfig struct {
 	//
 	// By default, there is no initial field.
-	InitialFields map[string]interface{} `mapstructure:"initial_fields"`
+	InitialFields map[string]interface{} `mapstructure:"initial_fields,omitempty"`
```
same

bogdandrutu commented on confmap/internal/mapstructure/encoder.go:
```diff
@@ -0,0 +1,234 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mapstructure // import "go.opentelemetry.io/collector/confmap/internal/mapstructure"
```
I was expected this to be a Decode into `map[string]interface{}`. Can you explain why the whole thing is needed?

jefchien replied:
`mapstructure.Decode` into `map[string]interface{}` doesn't support the mapstructure tags of nested structs. It'll only marshal the top level string key and then since the output map value is an interface, it'll just set the value (https://github.com/mitchellh/mapstructure/blob/bf980b35cac4dfd34e05254ee5aba086504c3f96/mapstructure.go#L509). It decodes based on the structure of the `result` in the decoder config.

I think `mapstructure` in general isn't really designed for encoding.

bogdandrutu replied:
Not convinced based on https://github.com/mitchellh/mapstructure/issues/53. Are you sure, can you write this to that issue with a test to prove?

bogdandrutu replied:
I understand why that does not work. I think we can have a "DecodeHook" that if the "from" is struct then it generates from to be map[string]interface I think that may work.

What do you think?

jefchien replied:
I think the main issue is more that the "to" reflect.Value is the bottleneck. It won't continue to run decode against the nested structs unless there's a corresponding "to" to decode into. If the "to" is just an interface, then it'll stop there. Unless we can modify the "to" in the "DecodeHook", I don't think a "DecodeHook" will help. I'll put an example in the issue you linked above when I get the chance.

jefchien replied:
I was a bit mistaken. After a bit more testing it turns out that it's actually the `decodeMap` function that won't work well with the `TextMarshalerHookFunc`. Turns out it doesn't call `decode` on the map keys for nested fields if it is marshalling a struct into a map, so the hook doesn't get called (https://github.com/mitchellh/mapstructure/blob/bf980b35cac4dfd34e05254ee5aba086504c3f96/mapstructure.go#L1006). This is a problem for us because the `service.Config` is just a bunch of maps with non-string (`ComponentID`) keys that need to be marshalled.

https://go.dev/play/p/0YBJAW-iCtP

It also turns out that running `Decode` with a struct that has a marshaller that marshals it into a string will break the `mapstructure` decoder. It looks like the other ticket is closed, but I'll put the findings in it as well.

jefchien replied:
I think this is more of the issue https://github.com/mitchellh/mapstructure/issues/166

jefchien replied:
Doesn't look like the mapstructure repo is very active. Don't think I'll be getting a response soon. Would it be possible to proceed with this workaround and then if the mapstructure repo is able to address the issue, replace the implementation? The implementation isn't exposed to the user, so there shouldn't be any issues with future backwards compatibility.
